osu.Game.Rulesets.Osu/Difficulty/Skills/Tap.cs

```diff
@@ -0,0 +1,161 @@
+            (List<OsuHitObject> hitObjects, double clockRate)
+        {
+            (var strainHistory, var tapDiff) = calculateTapStrain(hitObjects, 0, clockRate);
+            double burstStrain = strainHistory.Max(v => v[0]);
+
+            var streamnessMask = CalculateStreamnessMask(hitObjects, burstStrain, clockRate);
+            double streamNoteCount = streamnessMask.Sum();
+
+            (var mashLevels, var tapSkills) = calculateMashLevelsVSTapSkills(hitObjects, clockRate);
+
+            return (tapDiff, streamNoteCount, mashLevels, tapSkills, strainHistory);
+        }
+
+        /// <summary>
+        /// Calculates the strain values at each note and the maximum strain values
+        /// </summary>
+        private static (List<Vector<double>>, double) calculateTapStrain(List<OsuHitObject> hitObjects,
+                                                                                 double mashLevel,
+                                                                                 double clockRate)
+        {
+            var strainHistory = new List<Vector<double>> { decay_coeffs * 0, decay_coeffs * 0 };
+            var currStrain = decay_coeffs * 1;
+
+            // compute strain at each object and store the results into strainHistory
+            if (hitObjects.Count >= 2)
+            {
+                double prevPrevTime = hitObjects[0].StartTime / 1000.0;
+                double prevTime = hitObjects[1].StartTime / 1000.0;
+
+                for (int i = 2; i < hitObjects.Count; i++)
+                {
+                    double currTime = hitObjects[i].StartTime / 1000.0;
+
+                    // compute current strain after decay
+                    currStrain = currStrain.PointwiseMultiply((-decay_coeffs * (currTime - prevTime) / clockRate).PointwiseExp());
+
+                    strainHistory.Add(currStrain.PointwisePower(1.1 / 3) * 1.5);
+
+                    double distance = (hitObjects[i].Position - hitObjects[i - 1].Position).Length / (2 * hitObjects[i].Radius);
+                    double spacedBuff = calculateSpacedness(distance) * spaced_buff_factor;
+
+                    double deltaTime = Math.Max((currTime - prevPrevTime) / clockRate, 0.01);
+
+                    // for 1/4 notes above 200 bpm the exponent is -2.7, otherwise it's -2
+                    double strainAddition = Math.Max(Math.Pow(deltaTime, -2.7) * 0.265, Math.Pow(deltaTime, -2));
+
+                    currStrain += decay_coeffs * strainAddition *
+                                  Math.Pow(calculateMashNerfFactor(distance, mashLevel), 3) *
+                                  Math.Pow(1 + spacedBuff, 3);
+
+                    prevPrevTime = prevTime;
+                    prevTime = currTime;
+                }
+            }
+
+            // compute difficulty by aggregating strainHistory
+            var strainResult = decay_coeffs * 0;
```

Comment:
isn't this just 0?

Reply:
it's a new vector of the same shape as `decay_coeffs`

Reply:
might communicate the intent better by using `Vector<double>.Build(decay_coeffs.Count)`, or extract it out to a method like `NewOfShape(Vector v)` if it's something you do often. Unless there are performance impacts with building?

Reply:
That is really confusing to say the least, vector multiplication to create an identically sized empty vector does not really express intent well as @tybug said

Reply:
ok i'll change that